components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt

```diff
@@ -0,0 +1,541 @@
+            override fun toString(): String {
+                return "Failed to check MP: $e"
+            }
+        }
+
+        internal data class UnsupportedSignonsDbVersion(val version: Int) : Failure() {
+            override fun toString(): String {
+                return "Unsupported signons.sqlite DB version: $version"
+            }
+        }
+
+        internal data class UnexpectedLoginsKeyMaterialAlg(val berDecoded: String) : Failure() {
+            override fun toString(): String {
+                return "Unexpected login key material encryption alg oid: $berDecoded"
+            }
+        }
+
+        internal data class UnexpectedMetadataKeyMaterialAlg(val berDecoded: String) : Failure() {
+            override fun toString(): String {
+                return "Unexpected metadata key material encryption alg oid: $berDecoded"
+            }
+        }
+    }
+}
+
+@Suppress("TooManyFunctions", "LargeClass")
+internal object FennecLoginsMigration {
+    const val DEFAULT_MASTER_PASSWORD = ""
+    private const val EXPECTED_DB_VERSION = 6
+
+    private val logger = Logger("FennecLoginsMigration")
+    private val keyForId: MutableMap<String, ByteArray> = mutableMapOf()
+
+    @Suppress("TooGenericExceptionCaught", "ReturnCount")
+    internal fun migrate(
+        crashReporter: CrashReporter,
+        signonsDbPath: String,
+        key4DbPath: String,
+        masterPassword: String = DEFAULT_MASTER_PASSWORD
+    ): Result<LoginsMigrationResult> {
+        val noMasterPassword = try {
+            isMasterPasswordValid(masterPassword, key4DbPath)
+        } catch (e: Exception) {
+            logger.error("Failed to check master password", e)
+            return Result.Failure(LoginMigrationException(LoginsMigrationResult.Failure.FailedToCheckMasterPassword(e)))
+        }
+
+        if (!noMasterPassword) {
+            return Result.Success(LoginsMigrationResult.Success.MasterPasswordIsSet)
+        }
+
+        val loginRecords = try {
+            getLogins(crashReporter, masterPassword, signonsDbPath = signonsDbPath, key4DbPath = key4DbPath)
+        } catch (e: Exception) {
+            return Result.Failure(e)
+        }
+
+        return Result.Success(loginRecords)
+    }
+
+    /**
+     * Checks if provided [masterPassword] is valid (can be used to decrypt records).
+     *
+     * @param masterPassword A master password to check.
+     * @param key4DbPath Path to key4 db.
+     * @return True if [masterPassword] is correct, false otherwise.
+     */
+    internal fun isMasterPasswordValid(masterPassword: String, key4DbPath: String): Boolean {
+        val db = SQLiteDatabase.openDatabase(key4DbPath, null, OPEN_READONLY)
+        return db.rawQuery("select item1, item2 from metadata where id = 'password'", null).use {
+            it.moveToNext()
+
+            val globalSalt = it.getBlob(it.getColumnIndexOrThrow("item1"))
+            val item2 = it.getBlob(it.getColumnIndexOrThrow("item2"))
```

Comment:
Can we add the details from your comment https://github.com/mozilla-mobile/android-components/issues/4067#issuecomment-558424220 to describe what item2 is?